src/libraries/System.Security.Cryptography.Algorithms/src/Internal/Cryptography/Pbkdf2Implementation.Windows.cs

```diff
@@ -0,0 +1,265 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Diagnostics;
+using System.Security.Cryptography;
+using Microsoft.Win32.SafeHandles;
+using BCryptAlgPseudoHandle = Interop.BCrypt.BCryptAlgPseudoHandle;
+using BCryptBuffer = Interop.BCrypt.BCryptBuffer;
+using BCryptOpenAlgorithmProviderFlags = Interop.BCrypt.BCryptOpenAlgorithmProviderFlags;
+using NCryptBCryptBufferDescriptors = Interop.BCrypt.NCryptBCryptBufferDescriptors;
+using NTSTATUS = Interop.BCrypt.NTSTATUS;
+
+namespace Internal.Cryptography
+{
+    internal partial class Pbkdf2Implementation
+    {
+        private static readonly bool s_usePseudoHandles = OperatingSystem.IsWindowsVersionAtLeast(10, 0, 0);
+
+        // For Windows 7 we will use BCryptDeriveKeyPBKDF2. For Windows 8 we will use BCryptDeriveKey
+        // since it has better performance.
+        private static readonly bool s_useKeyDerivation = OperatingSystem.IsWindowsVersionAtLeast(8, 0, 0);
+
+        private static volatile SafeBCryptAlgorithmHandle? s_pbkdf2AlgorithmHandle;
+
+        public static unsafe void Fill(
+            ReadOnlySpan<byte> password,
+            ReadOnlySpan<byte> salt,
+            int iterations,
+            HashAlgorithmName hashAlgorithmName,
+            Span<byte> destination)
+        {
+            Debug.Assert(!destination.IsEmpty);
+            Debug.Assert(iterations >= 0);
+            Debug.Assert(hashAlgorithmName.Name is not null);
+
+            if (s_useKeyDerivation)
+            {
+                FillKeyDerivation(password, salt, iterations, hashAlgorithmName.Name, destination);
+            }
+            else
+            {
+                FillDeriveKeyPBKDF2(password, salt, iterations, hashAlgorithmName.Name, destination);
+            }
+        }
+
+        private static unsafe void FillKeyDerivation(
+            ReadOnlySpan<byte> password,
+            ReadOnlySpan<byte> salt,
+            int iterations,
+            string hashAlgorithmName,
+            Span<byte> destination)
+        {
+            SafeBCryptKeyHandle keyHandle;
+            int hashBlockSizeBytes = GetHashBlockSize(hashAlgorithmName);
+
+            // stackalloc 0 to let compiler know this cannot escape.
+            ReadOnlySpan<byte> symmetricKeyMaterial = stackalloc byte[0];
+            int symmetricKeyMaterialLength;
+            Span<byte> clearSpan = stackalloc byte[0];
+
+            if (password.IsEmpty)
+            {
+                symmetricKeyMaterial = stackalloc byte[1];
+                symmetricKeyMaterialLength = 0;
+                clearSpan = default;
+            }
+            else if (password.Length <= hashBlockSizeBytes)
```

Comment:
We _could_ let `BCryptGenerateSymmetricKey` do this for us, but it will only do it up to 2048 bytes before we need to manually hash the input anyway. I don't know if we gain much by letting `BCryptGenerateSymmetricKey` for values > block_size && < 2048 instead of just doing it ourselves with the one-shots. I'll put some performance numbers together, if the additional complexity is worth the gain, I can bring it in.